### DISCUSSION TITLE
Thinking about the Embedding Process

### DISCUSSION THREAD

**Discussion by hittjwX** [brianpetro/obsidian-smart-connections#745] [Ideas]
TL;DR; Would predigesting or embedding indexes improve fidelity of AI response? Would these non-note specific embeds improve relevance?

Would there be higher fidelity in AI responses if notes were predigested with timeline, machine learning contexts like Jaccard or Overlap, and natural language processing?

For example, before embedding, a map of notes modified and created in time, as well as the YAML of each note, is created. Then, someone can query, "List a summary of all notes in the last two days."

The YAML `type: Article` and `last modified <= 48 hours` would be embedded with the specific context of timeline and classification. This would give the AI more context about where to look.

Another example would be to extract all proper names, places, and things, classify them, and embed that classification. A prompt can ask, "What notes have I taken that would interest people like John Doe?" 

AI might respond as it does when asked to write in someone else's voice. This would help a marketer take notes about conversations with specific people and predict other topics worth future discussions.

As a newsletter publisher, I'd create a buyer persona around `John Doe`, link in some real customer conversations, and use geographic associations. "John Doe is a 28-year-old Midwestern male interested in hiking and technology. He is similar to [[Jane Doe]], [[Mike Doe]], and [[Roger Doe]] in outlook on life."

All my conversations with those other people, as well as any profiles and preferences, would be associated with `John Doe.` The pre-digestion may reinforce context even if I haven't had a specific conversation with the proxy.

I build contextual mapping using Dataview, DB Folder, and Graph Analysis. If those derived tables could be embedded, AI may reply in more detail. Fidelity, to me, is higher context and accuracy in predictions.

It's okay if this happens in the Smart Connect application, but it has to happen on my machine locally and securely. Smart actions aren't as important to me as improving the context between ideas and individuals.

**Comment by vdruts** [comment]
Fascinating idea. We're definitely barely scratching the surface of linking models to add context, and better mapping.

**Reply by philk27** [reply]
Methinks your asking for two things here (and please correct me if I'm wrong). Graph embeddings alongside vectors which enhances Indirect Relationship Discovery, Auto-detect clusters and automatically groups notes into thematic clusters for better Community Detection, Role-Based Recommendations, Dynamic Priority Weighting, Bridge Suggestion by identifying notes that connect disparate clusters.  The RAGFLOW project has that capability now in its knowledge base construction by way of reference or example.  

The second is splitting out tasks that are currently being handled by LLM in many implementations but would be better served by using ML and or neural network workflows.  

A lot of tasks around note organisation (and constructing good graph databases) hinges around categorisation and tagging – a cognitively intensive task that humans are not well suited for. The human brain is a great heuristic decision engine, a fair to middling programmatic/deterministic decision engine, and a terrible memory engine. Having machine language assisted guidance on generating, and using categories and tags (preferably based on existing taxonomies) is invaluable, especially for anyone in the Neurodivergent camp. Remembering what the categorisations are, the criteria for applying one, and applying them consistently as a task human brains are not good at, but AI assisted 2nd brains would be brilliant at.  That's just one example where splitting out ML/NN tasks, would be a huge benefit.

In saying that there's been a recent paper published of a co-embedding model  CEN-DGCNN, (Co-embedding of Edges and Nodes with Deep Graph Convolutional Neural Networks) which is a novel graph neural network architecture that simultaneously learns node and edge embeddings. It's ability to jointly model content (via quasi vector embeddings) and context (via graph embeddings) will be fantastic for hybrid search, knowledge graph construction and it's integration potential for complex document analysis use cases. That particular implementation does need a little bit of local power (at least an RTX 3080) but its implementation could be split out into smaller sequential processes that individually can be run quite parsimoniously.

We live in interesting times, but I'm seeing exactly the same patterns of proliferation followed by consolidation that we saw during desktop Network, Internet, cloud services adoption cycles.  This team commitment to building systems to defragment some of the ecosystem is great. Everybody using an API first methodology would be even better (would save me from having to build custom creations from different systems and databases into 1/2 decent second brain 🤭)

**Reply by hittjwX** [reply]
RAGFLOW is very interesting. I essentially wanted embeds to be smarter. To consume fewer resources, while deriving a base understanding of notes to create supplemental embeds. Great points all around.
